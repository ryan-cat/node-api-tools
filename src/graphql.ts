export interface IViewer {
  token: any;
  rawToken: string;
}

export interface Context {
  req: any;
  viewer: IViewer;
}
