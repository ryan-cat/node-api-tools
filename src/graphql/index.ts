export interface IViewer {
  token: any;
  rawToken: string;
}

export interface GraphQLContext {
  req: any;
  viewer: IViewer;
}
