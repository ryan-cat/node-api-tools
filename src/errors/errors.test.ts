import { ApolloError, isInstance } from 'apollo-errors';
import { GraphQLError } from 'graphql';
import {
  formatError,
  UserInputError,
  ValidationError,
  InternalServerError,
  ItemNotFoundError,
  AuthenticationError,
  AuthorizationError,
  AuthorizationActionError,
  APIError
} from './index';
import { ApolloServer } from 'apollo-server';

describe('errors', () => {
  it('should be instances of ApolloError', () => {
    expect(isInstance(new UserInputError(''))).toBeTruthy();
    expect(isInstance(new AuthenticationError())).toBeTruthy();
    expect(isInstance(new ItemNotFoundError(''))).toBeTruthy();
    expect(isInstance(new ValidationError(''))).toBeTruthy();
    expect(isInstance(new AuthorizationError(''))).toBeTruthy();
    expect(isInstance(new AuthorizationActionError(''))).toBeTruthy();
    expect(isInstance(new APIError(''))).toBeTruthy();
    expect(isInstance(new InternalServerError(''))).toBeTruthy();
  });

  it('should have correct message for user input error', () => {
    const message = 'Sample message';
    expect(new UserInputError(message).message).toMatchSnapshot();
  });

  it('should have correct message for item not found error for item and action', () => {
    const item = 'user';
    const action = 'create';
    expect(new ItemNotFoundError(item, action).message).toMatchSnapshot();
  });

  it('should have correct message for item not found error for item and no action', () => {
    const item = 'user';
    expect(new ItemNotFoundError(item).message).toMatchSnapshot();
  });

  it('should have correct message and data for validation error for item and no action', () => {
    const item = 'user';
    const error = new ValidationError(item);

    expect(error.message).toMatchSnapshot();
  });

  it('should have correct message and data for validation error for item and action and data', () => {
    const item = 'user';
    const action = 'create';
    const data = [{
      field: ['name'],
      error: 'error'
    }];
    const error = new ValidationError(item, action, data);

    expect(error.message).toMatchSnapshot();
    expect(error.data).toStrictEqual(data);
  });

  it('should have correct message for authorization error', () => {
    const message = 'Sample message';
    expect(new AuthorizationError(message).message).toMatchSnapshot();
  });

  it('should have correct message for authorization action error', () => {
    const action = 'create';
    expect(new AuthorizationActionError(action).message).toMatchSnapshot();
  });

  it('should have correct message for api error', () => {
    const message = 'Sample message';
    expect(new APIError(message).message).toMatchSnapshot();
  });
});

describe('format error', () => {
  it('should work with apollo server', () => {
    new ApolloServer({
      formatError: (err: GraphQLError) => formatError(err),
      typeDefs: `
        type Query {
          hello: String!
        }
      `,
      resolvers: {
        Query: {
          hello: () => {
            return 'world';
          }
        }
      }
    });
  });

  it('should not show original error when not type of ApolloError and masking', () => {
    const originalError = new Error('Error');

    const error = new GraphQLError('Something happened', null, null, null, null, originalError);
    const formattedError = formatError(error, true);

    expect(formattedError.message).toBe(new InternalServerError().message);
  });

  it('should show original error when type of ApolloError and masking', () => {
    const message = 'original error';
    const data = {
      field: 'name'
    };

    const originalError = new ApolloError('Error', { message, data }, { message, data });

    const error = new GraphQLError('Something happened', null, null, null, null, originalError);
    const formattedError = formatError(error, true);

    expect(formattedError.message).toBe(message);
    expect(formattedError.extensions.code).toBe(originalError.name);
    expect(formattedError.extensions.data).toStrictEqual(data);
  });

  it('should show original error when type of ApolloError and not masking', () => {
    const message = 'original error';
    const data = {
      field: 'name'
    };

    const originalError = new ApolloError('Error', { message, data }, { message, data });

    const error = new GraphQLError('Something happened', null, null, null, null, originalError);
    const formattedError = formatError(error, true);

    expect(formattedError.message).toBe(message);
    expect(formattedError.extensions.code).toBe(originalError.name);
    expect(formattedError.extensions.data).toStrictEqual(data);
  });

  it('should show original error when not type of ApolloError and not masking', () => {
    const message = 'original error';
    const originalError = new Error(message);

    const error = new GraphQLError('Something happened', null, null, null, null, originalError);
    const formattedError = formatError(error, false);

    expect(formattedError.message).toBe(message);
  });
});
