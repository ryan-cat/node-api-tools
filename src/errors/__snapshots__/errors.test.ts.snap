// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`errors should have correct message and data for validation error for item and action and data 1`] = `"user could not be create due to issues with the provided input."`;

exports[`errors should have correct message and data for validation error for item and no action 1`] = `"user could not be saved due to issues with the provided input."`;

exports[`errors should have correct message for api error 1`] = `"Sample message"`;

exports[`errors should have correct message for authorization action error 1`] = `"You do not have permission to create."`;

exports[`errors should have correct message for authorization error 1`] = `"Sample message"`;

exports[`errors should have correct message for item not found error for item and action 1`] = `"The user could not be found to create."`;

exports[`errors should have correct message for item not found error for item and no action 1`] = `"The user could not be found."`;

exports[`errors should have correct message for user input error 1`] = `"Sample message"`;
